include/matrix3x4SIMD.h

```diff
@@ -306,55 +317,58 @@ struct matrix3x4SIMD
-		vectorSIMDf tmpC = tmpA - tmpB; //1st column of _out matrix
-		vectorSIMDf preDeterminant = rows[0] * tmpC;
-		preDeterminant = _mm_hadd_ps(preDeterminant.getAsRegister(), preDeterminant.getAsRegister()); // (x+y,z+w,..)
-		vectorSIMDf determinant4 = _mm_hadd_ps(preDeterminant.getAsRegister(), preDeterminant.getAsRegister());
+		vectorSIMDf c0 = rows[0], c1 = rows[1], c2 = rows[2], c3 = vectorSIMDf(0.f, 0.f, 0.f, 1.f);
+		core::transpose4(c0, c1, c2, c3);
 
-		if (determinant4.x == 0.f)
+		const vectorSIMDf c1crossc2 = c1.crossProduct(c2);
+
+		const vectorSIMDf d = c0.dotProduct(c1crossc2);
+
+		if (core::iszero(d.x, FLT_MIN))
 			return false;
 
-		tmpA = rows[0].zxyw()*rows[2].yzxw();
-		tmpB = rows[0].yzxw()*rows[2].zxyw();
-		vectorSIMDf tmpD = tmpA - tmpB; // 2nd column of _out matrix
+		_out.rows[0] = c1crossc2 / d;
+		_out.rows[1] = (c2.crossProduct(c0)) / d;
+		_out.rows[2] = (c0.crossProduct(c1)) / d;
 
-		tmpA = rows[0].yzxw()*rows[1].zxyw();
-		tmpB = rows[0].zxyw()*rows[1].yzxw();
-		vectorSIMDf tmpE = tmpA - tmpB; // 3rd column of _out matrix
+		vectorSIMDf outC3 = vectorSIMDf(0.f, 0.f, 0.f, 1.f);
+		core::transpose4(_out.rows[0], _out.rows[1], _out.rows[2], outC3);
+		mulSub3x3With3x1(outC3.pointer, c3.pointer);
+		outC3 = -outC3;
+		core::transpose4(_out.rows[0], _out.rows[1], _out.rows[2], outC3);
 
-		__m128 xmm0 = tmpC.getAsRegister();
-		__m128 xmm1 = tmpD.getAsRegister();
-		__m128 xmm2 = tmpE.getAsRegister();
-		__m128 xmm3 = _mm_setzero_ps();
+		return true;
+	}
+	bool makeInverse()
+	{
+		matrix3x4SIMD tmp;
 
-		_MM_TRANSPOSE4_PS(xmm0, xmm1, xmm2, xmm3);
+		if (getInverse(tmp))
+		{
+			*this = tmp;
+			return true;
+		}
+		return false;
+	}
 
-		__m128 xmm4 = getTranslation3D().getAsRegister();
+	inline bool getSub3x3Inverse(float* _out) const
+	{
+		vectorSIMDf c0 = rows[0], c1 = rows[1], c2 = rows[2], c3 = vectorSIMDf(0.f, 0.f, 0.f, 1.f);
+		core::transpose4(c0, c1, c2, c3);
 
-		xmm0 = _mm_mul_ps(xmm0, xmm4); //_out(0,3)
-		xmm1 = _mm_mul_ps(xmm1, xmm4); //_out(1,3)
-		xmm2 = _mm_or_ps(_mm_mul_ps(xmm2, xmm4), _mm_castsi128_ps(_mm_set_epi32(0, -1, 0, -1))); //_out(2,3)
+		const vectorSIMDf c1crossc2 = c1.crossProduct(c2);
 
-		xmm0 = _mm_hsub_ps(xmm0, xmm1); // C.x-D.x,E.x,C.y-D.y,E.y
-		xmm1 = _mm_hsub_ps(xmm2, preDeterminant.getAsRegister()); // C.z-D.z,E.z,x+y-z-w,x+Y-z-w
-		xmm2 = _mm_hsub_ps(xmm0, xmm1); // C.x-D.x-E.x,C.y-D.y-E.y,C.z-D.z-E.z,0
+		const vectorSIMDf d = c0.dotProduct(c1crossc2);
 
-		__m128 tc = tmpC.getAsRegister(), td = tmpD.getAsRegister(), te = tmpE.getAsRegister();
-		_MM_TRANSPOSE4_PS(tc, td, td, xmm2);
-		_out.rows[0] = tc;
-		_out.rows[1] = td;
-		_out.rows[2] = te;
+		if (core::iszero(d.x, FLT_MIN))
+			return false;
 
-		__m128 rdet = _mm_rcp_ps(determinant4.getAsRegister());
-		_out.rows[0] *= rdet;
-		_out.rows[1] *= rdet;
-		_out.rows[2] *= rdet;
+		reinterpret_cast<vectorSIMDf*>(_out)[0] = c1crossc2 / d;
+		reinterpret_cast<vectorSIMDf*>(_out+3)[0] = (c2.crossProduct(c0)) / d;
```

Comment:
this will crash, vectorSIMDf has a memory alignmend requirement of 16 bytes, plus you can't guarantee _out is 16 byte aligned in the first place